### ISSUE TITLE
SNAT traffic from outside-cluster sources in kube-proxy

### ISSUE THREAD

**Issue by cjcullen** [kubernetes/kubernetes#24224]
A lot of users want to access kubernetes services from outside the cluster, but on the same private network. With the userspace kube-proxy, there are a couple of hacks that could make this work:

1. Run kube-proxy wherever you want to access those services.
2. Create bastion route(s) to funnel traffic to services to a cluster node.

The iptables kube-proxy breaks option 2. Traffic from a client that is bastioned to node A, but kube-proxied to a Pod on node B gets a response directly to the client, who is expecting to see a response from the service IP.

One workaround is to masquerade traffic from IPs outside of the cluster. It means that we don't see true client IPs, but that was the case with the bastion route w/ the userspace proxy anyway.

So, could we add a flag to kube-proxy that gets it to maintain a masquerade rule for off-cluster traffic? Something like:
```
iptables -t nat -I POSTROUTING ! -s "${CLUSTER_CIDR}" -j MASQUERADE
```

**Comment by cjcullen** [comment]
@ArtfulCoder @thockin @Q-Lee

**Comment by thockin** [comment]
For the record, I dislike using Cluster CIDR in this way (I dislike the
notion of cluster CIDR) but I don't have a better way to express "traffic
from off machine" in a generic way.  Ideas welcome.

I'd like to get this in for the next patch release, if possible, since it
used to work (even if it was undocumented).

**Comment by goblain** [comment]
@cjcullen : If I follow you correctly, in this situation you do have a full connectivity to pods on your node (like ie. overlay networking in flannel) so that local iptables kube-proxy can DNAT to POD IP. So this is similar to a case when you wont to access POD from node/minion it self (not from withing the container).

I had this problem on my CoreOS cluster with flannel. The core of mine issue was not the fact that the response comes from pod IP, but that when you send traffic to service IP it's destination IP is not a part of flannel0 (or other overlay etc.), hence it goes out via default routes interce, and recieves a source IP of your main interface like eth0. Then kube-proxy rules kick in, change destination and send the traffic to target pod. Unfortunately the pod responds to your source IP so it sends completely different way it should. The rule you suggest above fixes this issue indeed, but I ended up with slightly different one :
```
/sbin/iptables -t nat -I POSTROUTING -o flannel0 -s $private_ipv4 -j MASQUERADE
```
as this is used from CloudFormation the $private_ipv4 is my eth0 IP. The problem with this rule is that it will not catch services that are run on the host in non-k8s docker, vms etc. so `! -s ${CLUSTER_CIDR}` might still be a better approach, but I would stick to explicit use of the outgoing interface so that we do not cover too much with the MASQUERADE (like ie. all the traffic going from routed VMs on host to the outside world).

**Comment by thockin** [comment]
CJ - is this something you're going to have time to do, or are you looking at me while I look at you?

**Comment by sbadia** [comment]
For the record,

Using the pure iptables kube-proxy, with the option `--masquerade-all=true` it seems to work well. I just tested this on GCE (Container clusters v1.2.1).
##### Create a bastion route on any node of the cluster
```
gcloud compute routes create svc-foo1 --network foo1 --destination-range x.x.x.x/20 --next-hop-instance gke-foobarnode1-cluster1
```
##### On cluster nodes, enable `masquerade-all=true`
```bash
# Kill running kube-proxy
kubeproxy_container=$(docker ps --filter=name=kube-proxy --filter=status=running --filter=label=io.kubernetes.container.name=kube-proxy --format "{{.ID}}")
docker kill $kubeproxy_container ; docker rm $kubeproxy_container
# Enable the option in the kubelet config.
sed "s|--v=2|--v=2 --masquerade-all=true|g" -i /etc/kubernetes/manifests/kube-proxy.manifest
# Restart
/etc/init.d/kubelet restart
```

**Comment by zhouhaibing089** [comment]
+1 for `--masquerade-all=true`.

my question maybe a little bit off-topic, not sure why it is not default to be `true`? in our environment, it has to be true.

if there is no snat, the packet could be received correctly, but the source ip is still being pod ip.

**Comment by goblain** [comment]
@zhouhaibing089 in case you are using networking which goes via plain routing (ie. via cloud provider routing) then --masquerade-all is not the best idea. Even though the traffic goes via "external" router, this router knows how to route CLUSTER_CIDR subnets hence you want to keep source pod IP unaltered. The SNAT in such case has to happen either when trafic leaves your cluster via gateway that will perform NAT or in case of nodes that can communicate with the world directly, when the target is not in CLUSTER_CIDR

**Comment by cjcullen** [comment]
@thockin I'll take a look today.

**Comment by cjcullen** [comment]
@goblain: I'm not familiar with how this would work on flannel. Maybe @thockin knows more.

The problem I'm trying to fix is traffic from off-cluster. If we only DNAT (serviceIP->podIP), then the response to the client will come from the podIP, when it is expecting a response from serviceIP.

**Comment by thockin** [comment]
For flannel we need the bastion to be on the flannel network or to route
packets to a machine configured for bridging, which i think flannel does by
default.

**Comment by cjcullen** [comment]
Fixed by #24429
